### ISSUE TITLE
Divergence between `noqa` and end-of-line `ruff:ignore`

### ISSUE THREAD

**Issue by ntBre** [astral-sh/ruff#25644]
I was working on a prototype of an `--add-ignore` flag to add `ruff:ignore` comments instead of `noqa` comments like `--add-noqa`, and I had finished a rough draft when Codex pointed out that my implementation was broken because it assumed `noqa` and `ruff:ignore` had the same semantics. I skimmed through our previous discussion on https://github.com/astral-sh/ruff/pull/23404 and don't _think_ this was intentional because it means `ruff:ignore` diverges both from `noqa` and from `ty:ignore`. This was the example Codex showed:

```py
suppressed = [  # noqa: RUF015
    *range(10)
][0]

not_suppressed = [  # ruff:ignore[RUF015]
    *range(10)
][0]

```

And a similar case in ty, which is also suppressed:

```py
value: int = (  # ty:ignore[invalid-assignment]
    "hello"
)
```

[Ruff Playground](https://play.ruff.rs/9d0d7d59-0502-465d-984e-a3f9d57564e8)

[ty Playground](https://play.ty.dev/f659f5e9-2319-4173-953e-24e84c2d263f)

Assuming this was not intentional, I'll try to fix this and then nearly all that needs to change for `--add-ignore` is the final text that gets written.

**Comment by MichaReiser** [comment]
I'm not a 100% sure but I believe we set a special range for 

```py
value: int = (  # ty:ignore[invalid-assignment]
    "hello"
)
``` 

in ty to make this work (the range is the parenthesized expression and not just `"hello"`


But I would expect

```py
not_suppressed = [  # ruff:ignore[RUF015]
    *range(10)
][0]
```

to work (https://play.ruff.rs/691df2f9-3eb7-40a6-a13e-875cc13a0bfb). Is it because ruff: ignore only allows suppression at the end line and not the start line?

**Comment by ntBre** [comment]
The highlighted range for this diagnostic, at least in the playground, is from the opening `[` all the way to the end of the slice. I think the problem is that the `ruff:ignore` suppression checks that its range fully contains the diagnostic range, which isn't the case here:

https://github.com/astral-sh/ruff/blob/4ef420da4712d653b3124398b6d2cf519b0bbf7f/crates/ruff_linter/src/suppression.rs#L231-L233

I think `noqa` and `ty:ignore` just check if the diagnostic starts (or ends for ty) on the same line instead.
